share/spack/gitlab/cloud_pipelines/stacks/ml/spack.yaml

```diff
@@ -0,0 +1,136 @@
+spack:
+  view: false
+
+  concretizer:
+    reuse: false
+    unify: false
+
+  config:
+    concretizer: clingo
+    install_tree:
+      root: /home/software/spack
+      padded_length: 384
+      projections:
+        all: "{architecture}/{compiler.name}-{compiler.version}/{name}-{version}-{hash}"
+
+  packages:
+    all:
+      compiler: [gcc@11.2.0]
+      target: [x86_64_v4]
+      variants: +cuda cuda_arch=80
+
+  specs:
+    # Horovod
+    - py-horovod
+
+    # JAX
+    - py-jax
+    - py-jaxlib
+
+    # Keras
+    - py-keras
+    - py-keras-applications
+    - py-keras-preprocessing
+    - py-keras2onnx
+
+    # PyTorch
+    - py-botorch
+    - py-efficientnet-pytorch
+    - py-gpytorch
+    - py-kornia
+    - py-pytorch-gradual-warmup-lr
+    - py-pytorch-lightning
+    - py-segmentation-models-pytorch
+    - py-timm
+    - py-torch
+    - py-torch-cluster
+    - py-torch-scatter
+    - py-torch-sparse
+    - py-torch-spline-conv
+    - py-torchaudio
+    - py-torchdata
+    - py-torchfile
+    - py-torchgeo
+    - py-torchmeta
+    - py-torchmetrics
+    - py-torchtext
+    - py-torchvision
+    - py-vector-quantize-pytorch
+
+    # scikit-learn
+    - py-scikit-learn
+    - py-scikit-learn-extra
+
+    # TensorBoard
+    - py-tensorboard
+    - py-tensorboard-data-server
+    - py-tensorboard-plugin-wit
+    - py-tensorboardx
+
+    # TensorFlow
+    - py-tensorflow
+    - py-tensorflow-hub
+    - py-tensorflow-probability
+
+    # XGBoost
+    - py-xgboost
+    - xgboost
+
+  mirrors: { "mirror": "s3://spack-binaries/develop/ml" }
+
+  gitlab-ci:
+    script:
+      - . "./share/spack/setup-env.sh"
+      - spack --version
+      - cd ${SPACK_CONCRETE_ENV_DIR}
+      - spack env activate --without-view .
+      - spack config add "config:install_tree:projections:${SPACK_JOB_SPEC_PKG_NAME}:'morepadding/{architecture}/{compiler.name}-{compiler.version}/{name}-{version}-{hash}'"
+      - mkdir -p ${SPACK_ARTIFACTS_ROOT}/user_data
+      - if [[ -r /mnt/key/intermediate_ci_signing_key.gpg ]]; then spack gpg trust /mnt/key/intermediate_ci_signing_key.gpg; fi
+      - if [[ -r /mnt/key/spack_public_key.gpg ]]; then spack gpg trust /mnt/key/spack_public_key.gpg; fi
+      - spack -d ci rebuild > >(tee ${SPACK_ARTIFACTS_ROOT}/user_data/pipeline_out.txt) 2> >(tee ${SPACK_ARTIFACTS_ROOT}/user_data/pipeline_err.txt >&2)
+
+    image: ecpe4s/ubuntu22.04-runner-x86_64:2022-07-01
```

Comment:
I think this instance of `image` was added unintentionally?  Just basing that on the fact that you generated the pipeline in `.gitlab-ci.yml` using the amazonlinux image.  The schema probably shouldn't allow two at the same level like this, since not obvious which one will be picked.

Reply:
I'm gonna be honest with you, I have no idea what an image is, nor what this specific image is. Most of this file is just copied and pasted from other files. When something didn't work, I tried copying different things.

Should this line or the other line be modified? What should it be modified to?

Reply:
Sometimes people who add a stack want to make sure it builds on a specific operating system or a set of them.  The `image` key allows specifying which docker image (and thus which os, mostly) should be used for various jobs.  So if you really don't care whether these are built on amazon linux or ubuntu, I guess you could flip a coin, and then just pick the same one for concretization (defined in bits you added to `.gitlab-ci.yml`) and rebuilds (above).

Reply:
Think I got it right...